packages/react-widgets/src/widgets/TimeSeriesWidget.js

```diff
@@ -0,0 +1,308 @@
+
+  const filters = useSelector((state) => {
+    const source = selectSourceById(state, dataSource) || {};
+    const filters = { ...source.filters };
+    delete filters[column];
+    return filters;
+  });
+
+  const isSourceReady = !!viewportFeaturesReady[dataSource];
+
+  useEffect(() => {
+    if (isSourceReady) {
+      setIsLoading(true);
+      const _filters = getApplicableFilters(filters, id);
+
+      getTimeSeries({
+        filters: _filters,
+        dataSource,
+        column,
+        stepSize: selectedStepSize,
+        operationColumn,
+        operation
+      })
+        .then((data) => {
+          if (data) {
+            setIsLoading(false);
+            setTimeSeriesData(data);
+          }
+        })
+        .catch((error) => {
+          setIsLoading(false);
+          if (onError) onError(error);
+        });
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [
+    id,
+    // filters is always a new object, then we have to stringify it
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    JSON.stringify(filters),
+    dataSource,
+    column,
+    selectedStepSize,
+    isSourceReady,
+    setIsLoading,
+    operationColumn,
+    operation,
+    onError
+  ]);
+
+  const handleTimeframeUpdate = useCallback(
+    (timeframe) => {
+      dispatch(
+        addFilter({
+          id: dataSource,
+          column,
+          type: FilterTypes.BETWEEN,
+          values: [timeframe],
+          owner: id
+        })
+      );
+
+      if (onTimeframeUpdate) onTimeframeUpdate();
+    },
+    [column, dataSource, dispatch, id, onTimeframeUpdate]
+  );
+
+  const handleTimelineUpdate = useCallback(
+    (timelinePosition) => {
+      const { name: moment } = timeSeriesData[timelinePosition];
+      handleTimeframeUpdate([
+        moment,
+        moment + BUCKET_SIZE_RANGE_MAPPING[selectedStepSize] * 1000
+      ]);
+
+      if (onTimelineUpdate) onTimelineUpdate();
+    },
+    [handleTimeframeUpdate, onTimelineUpdate, selectedStepSize, timeSeriesData]
+  );
+
+  const handleStop = useCallback(() => {
+    dispatch(
+      removeFilter({
+        id: dataSource,
+        column
+      })
+    );
+
+    if (onStop) onStop();
+  }, [column, dataSource, dispatch, onStop]);
+
+  const [anchorEl, setAnchorEl] = useState(null);
+
+  const handleOpenBucketSizeMenu = (e) => {
+    if (e?.currentTarget) {
+      setAnchorEl(e.currentTarget);
+    }
+  };
+
+  const handleCloseBucketSizeMenu = () => {
+    setAnchorEl(null);
+  };
+
+  const handleStepSizeUpdate = (stepSize) => {
+    setSelectedStepSize(stepSize);
+    handleCloseBucketSizeMenu();
+  };
+
+  return (
+    <>
+      <WrapperWidgetUI
+        title={title}
+        isLoading={isLoading}
+        {...wrapperProps}
+        actions={[
+          ...(wrapperProps?.actions || []),
+          ...(stepSizeOptions?.length
+            ? [
+                {
+                  id: 'a0',
```

Comment:
why 'a0' here? is it relevant? is it used somehow outside? Thinking if this should be an exported CONST or not

Reply:
Hmm I copied that from WrapperWidgetUI example. What should it be?

```
{
                  id: 'bucketSize',
                  name: 'Bucket size',
                  icon: <StepSizeIcon />,
                  action: handleOpenBucketSizeMenu
}
```

Is it ok?